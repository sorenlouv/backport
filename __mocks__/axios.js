const axios = jest.genMockFromModule('axios');
axios.get = jest.fn(() => {
  throw new Error('Forgot to mock axios.get');
});
axios.post = jest.fn(() => {
  throw new Error('Forgot to mock axios.post');
});
module.exports = axios;
