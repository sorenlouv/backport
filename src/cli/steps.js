const github = require('../lib/github');
const {
  getCommitByPrompt,
  getCommitBySha,
  getBranchesByPrompt,
  doBackportVersions,
  handleErrors,
  maybeSetupRepo
} = require('./cliService');

function initSteps(options) {
  const [owner, repoName] = options.upstream.split('/');

  let commits, branches;
  github.setAccessToken(options.accessToken);

  const promise = options.sha
    ? getCommitBySha({ owner, repoName, sha: options.sha })
    : getCommitByPrompt({
        owner,
        repoName,
        author: options.own ? options.username : null,
        multipleCommits: options.multipleCommits
      });

  return promise
    .then(c => {
      commits = c;
    })
    .then(() => getBranchesByPrompt(options.branches, options.multipleBranches))
    .then(v => {
      branches = v;
    })
    .then(() => maybeSetupRepo(owner, repoName, options.username))
    .then(() =>
      doBackportVersions({
        owner,
        repoName,
        commits,
        branches,
        username: options.username,
        labels: options.labels
      })
    )
    .catch(handleErrors);
}

module.exports = initSteps;
