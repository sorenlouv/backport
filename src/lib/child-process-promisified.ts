import childProcess from 'child_process';
import { promisify } from 'util';
import { logger } from './logger';
const execPromisified = promisify(childProcess.exec);

export async function exec(
  cmd: string,
  options: childProcess.ExecOptions & { cwd: string }
) {
  const res = await execPromisified(cmd, {
    maxBuffer: 100 * 1024 * 1024,
    ...options,

    // ensure that git commands return english error messages
    env: { ...process.env, LANG: 'en_US' },
  });

  return res;
}

export async function spawnPromise(
  cmd: string,
  cmdArgs: string[],
  cwd: string
): Promise<{
  cmdArgs: string[];
  code: number | null;
  stderr: string;
  stdout: string;
}> {
  const fullCmd = `${cmd} ${cmdArgs.join(' ')}`;
  logger.info(`Running command: "${fullCmd}"`);

  return new Promise(function (resolve, reject) {
    const subprocess = childProcess.spawn(cmd, cmdArgs, {
      cwd,

      // ensure that git commands return english error messages
      env: { ...process.env, LANG: 'en_US' },
    });
    let stderr = '';
    let stdout = '';

    subprocess.stdout.on('data', (data: string) => {
      stdout += data;
    });

    subprocess.stderr.on('data', (data: string) => {
      stderr += data;
    });

    subprocess.on('close', (code) => {
      if (code === 0 || code === null) {
        resolve({ cmdArgs, code, stderr, stdout });
      } else {
        const err = new SpawnError({ cmdArgs, code, stderr, stdout });
        logger.verbose(`Error when running command: "${fullCmd}"`, err);
        reject(err);
      }
    });

    subprocess.on('error', (err) => {
      reject(err);
    });
  });
}

export const spawnStream = (cmd: string, cmdArgs: ReadonlyArray<string>) => {
  return childProcess.spawn(cmd, cmdArgs, {
    env: { ...process.env, LANG: 'en_US' },
  });
};

export type SpawnErrorContext = {
  cmdArgs: string[];
  code: number;
  stderr: string;
  stdout: string;
};

export class SpawnError extends Error {
  context: SpawnErrorContext;
  constructor(context: SpawnErrorContext) {
    const cmdArgs = context.cmdArgs.join(' ');
    const message = `Code: ${
      context.code
    }, Args: "${cmdArgs}", Message: ${context.stderr.trim()}`;

    super(message);
    Error.captureStackTrace(this, SpawnError);
    this.name = 'SpawnError';
    this.message = message;
    this.context = context;
  }
}
