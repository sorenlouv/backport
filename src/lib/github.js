const axios = require('axios');
const querystring = require('querystring');
const get = require('lodash.get');
const constants = require('./constants');

let accessToken;
function getCommitMessage(message) {
  return message.split('\n')[0].trim();
}

function getCommits(owner, repoName, author) {
  const urlArgs = {
    per_page: 20,
    access_token: accessToken
  };

  if (author) {
    urlArgs.author = author;
    urlArgs.per_page = 5;
  }

  return axios(
    `https://api.github.com/repos/${owner}/${repoName}/commits?${querystring.stringify(
      urlArgs
    )}`
  )
    .catch(handleError)
    .then(res =>
      res.data.map(commit => {
        const message = getCommitMessage(commit.commit.message);
        return {
          message,
          sha: commit.sha
        };
      })
    );
}

function getCommit(owner, repoName, sha) {
  return axios(
    `https://api.github.com/repos/${owner}/${repoName}/commits/${sha}?access_token=${accessToken}`
  )
    .catch(handleError)
    .then(res => ({
      message: res.data.commit.message,
      sha: res.data.sha
    }));
}

function createPullRequest(owner, repoName, payload) {
  return axios
    .post(
      `https://api.github.com/repos/${owner}/${repoName}/pulls?access_token=${accessToken}`,
      payload
    )
    .catch(handleError);
}

function addLabels(owner, repoName, pullNumber, labels) {
  return axios
    .post(
      `https://api.github.com/repos/${owner}/${repoName}/issues/${pullNumber}/labels?access_token=${accessToken}`,
      labels
    )
    .catch(handleError);
}

function getPullRequestByCommit(owner, repoName, commitSha) {
  return axios(
    `https://api.github.com/search/issues?q=repo:${owner}/${repoName}+${commitSha}&access_token=${accessToken}`
  )
    .catch(handleError)
    .then(res => get(res.data.items[0], 'number'));
}

function setAccessToken(_accessToken) {
  accessToken = _accessToken;
}

class GithubError extends Error {
  constructor(message) {
    super();
    Error.captureStackTrace(this, GithubError);
    this.code = constants.GITHUB_ERROR;
    this.message = message;
  }
}

function handleError(e) {
  if (get(e.response, 'data')) {
    throw new GithubError(e.response.data);
  }

  throw e;
}

module.exports = {
  setAccessToken,
  addLabels,
  createPullRequest,
  getCommit,
  getCommits,
  getPullRequestByCommit
};
