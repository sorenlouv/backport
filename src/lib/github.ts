import {
  GithubQuery,
  GithubCommit,
  PullRequest,
  Commit,
  GithubIssue,
  GithubPullRequestPayload,
  GithubIssues,
  GithubApiError
} from '../types/types';

import axios, { AxiosResponse } from 'axios';
import querystring from 'querystring';
import get from 'lodash.get';
import { HandledError } from './errors';

let accessToken: string;
function getCommitMessage(message: string) {
  return message.split('\n')[0].trim();
}

export async function getCommits(
  owner: string,
  repoName: string,
  author: string | null
): Promise<Commit[]> {
  const query: GithubQuery = {
    access_token: accessToken,
    per_page: 20
  };

  if (author) {
    query.author = author;
    query.per_page = 5;
  }

  try {
    const res: AxiosResponse<GithubCommit[]> = await axios(
      `https://api.github.com/repos/${owner}/${repoName}/commits?${querystring.stringify(
        query
      )}`
    );

    const promises = res.data.map(async commit => {
      const sha = commit.sha;
      return {
        message: getCommitMessage(commit.commit.message),
        sha,
        pullRequest: await getPullRequestBySha(owner, repoName, sha)
      };
    });

    return Promise.all(promises);
  } catch (e) {
    throw getError(e);
  }
}

export async function getCommit(
  owner: string,
  repoName: string,
  sha: string
): Promise<Commit> {
  try {
    const res: AxiosResponse<GithubCommit> = await axios(
      `https://api.github.com/repos/${owner}/${repoName}/commits/${sha}?access_token=${accessToken}`
    );
    const fullSha = res.data.sha;
    const pullRequest = await getPullRequestBySha(owner, repoName, fullSha);

    return {
      message: getCommitMessage(res.data.commit.message),
      sha: fullSha,
      pullRequest
    };
  } catch (e) {
    throw getError(e);
  }
}

export async function createPullRequest(
  owner: string,
  repoName: string,
  payload: GithubPullRequestPayload
): Promise<PullRequest> {
  try {
    const res: AxiosResponse<GithubIssue> = await axios.post(
      `https://api.github.com/repos/${owner}/${repoName}/pulls?access_token=${accessToken}`,
      payload
    );
    return {
      html_url: res.data.html_url,
      number: res.data.number
    };
  } catch (e) {
    throw getError(e);
  }
}

export async function addLabels(
  owner: string,
  repoName: string,
  pullNumber: number,
  labels: string[]
) {
  try {
    return await axios.post(
      `https://api.github.com/repos/${owner}/${repoName}/issues/${pullNumber}/labels?access_token=${accessToken}`,
      labels
    );
  } catch (e) {
    throw getError(e);
  }
}

async function getPullRequestBySha(
  owner: string,
  repoName: string,
  commitSha: string
): Promise<number> {
  try {
    const res: AxiosResponse<GithubIssues> = await axios(
      `https://api.github.com/search/issues?q=repo:${owner}/${repoName}+${commitSha}+base:master&access_token=${accessToken}`
    );
    return get(res.data.items[0], 'number');
  } catch (e) {
    throw getError(e);
  }
}

export function setAccessToken(token: string) {
  accessToken = token;
}

function getError(e: GithubApiError) {
  if (e.response && e.response.data) {
    return new HandledError(
      JSON.stringify({ ...e.response.data, axiosUrl: e.config.url }, null, 4)
    );
  }

  return e;
}
