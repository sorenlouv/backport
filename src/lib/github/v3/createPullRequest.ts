import { Octokit } from '@octokit/rest';
import apm from 'elastic-apm-node';
import Handlebars from 'handlebars';
import { ora } from '../../../lib/ora';
import { ValidConfigOptions } from '../../../options/options';
import { PACKAGE_VERSION } from '../../../utils/packageVersion';
import { BackportError } from '../../BackportError';
import { getSourceBranchFromCommits } from '../../getSourceBranchFromCommits';
import { logger } from '../../logger';
import { Commit } from '../../sourceCommit/parseSourceCommit';
import { getFirstLine, getShortSha } from '../commitFormatters';
import { fetchExistingPullRequest } from '../v4/fetchExistingPullRequest';
import { getGithubV3ErrorMessage } from './getGithubV3ErrorMessage';

export interface PullRequestPayload {
  owner: string;
  repo: string;
  title: string;
  body: string;
  head: string;
  base: string;
  [key: string]: unknown;
}

export async function createPullRequest({
  options,
  prPayload,
}: {
  options: ValidConfigOptions;
  prPayload: PullRequestPayload;
}): Promise<{
  url: string;
  number: number;
  didUpdate: boolean;
}> {
  logger.info(
    `Creating PR with title: "${prPayload.title}". ${prPayload.head} -> ${prPayload.base}`,
  );

  const { accessToken, githubApiBaseUrlV3 } = options;
  const spinner = ora(options.interactive, `Creating pull request`).start();

  if (options.dryRun) {
    spinner.succeed();
    return { didUpdate: false, number: 1337, url: 'this-is-a-dry-run' };
  }

  const span = apm.startSpan('REST: Create pull request');

  try {
    const octokit = new Octokit({
      auth: accessToken,
      baseUrl: githubApiBaseUrlV3,
      log: logger,
    });

    const res = await octokit.pulls.create(prPayload);

    spinner.succeed();

    return {
      url: res.data.html_url,
      number: res.data.number,
      didUpdate: false,
    };
  } catch (e) {
    // retrieve url for existing
    try {
      const existingPR = await fetchExistingPullRequest({
        options,
        prPayload,
      });

      if (existingPR) {
        spinner.succeed('Updating existing pull request');
        return {
          url: existingPR.url,
          number: existingPR.number,
          didUpdate: true,
        };
      }
    } catch (e) {
      logger.error('Could not retrieve existing pull request', e);
      // swallow error
    } finally {
      span?.end();
    }

    spinner.fail();
    throw new BackportError(
      //@ts-expect-error
      `Could not create pull request: ${getGithubV3ErrorMessage(e)}`,
    );
  }
}

export function getPullRequestBody({
  options,
  commits,
  targetBranch,
}: {
  options: ValidConfigOptions;
  commits: Commit[];
  targetBranch: string;
}) {
  const commitMessages = commits
    .map((c) => {
      const message = c.sourcePullRequest
        ? `[${getFirstLine(c.sourceCommit.message)}](${
            c.sourcePullRequest.url
          })`
        : `${getFirstLine(c.sourceCommit.message)} (${getShortSha(
            c.sourceCommit.sha,
          )})`;

      return ` - ${message}`;
    })
    .join('\n');

  const sourceBranch = getSourceBranchFromCommits(commits);

  const defaultPrDescription =
    '# Backport\n\n' +
    'This will backport the following commits from `{{sourceBranch}}` to `{{targetBranch}}`:\n' +
    '{{commitMessages}}\n\n' +
    '<!--- Backport version: {{PACKAGE_VERSION}} -->\n\n' +
    '### Questions ?\n' +
    'Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)';

  const customPrDescription = options.prDescription
    ?.replaceAll('{{defaultPrDescription}}', defaultPrDescription)
    ?.replaceAll('{defaultPrDescription}', defaultPrDescription) // retain backwards compatibility
    ?.replaceAll('{{commitsStringified}}', JSON.stringify(commits))
    ?.replaceAll('{commits}', JSON.stringify(commits)); // retain backwards compatibility

  const template = Handlebars.compile(
    customPrDescription ?? defaultPrDescription,
  );

  return template({
    sourcePullRequest: commits[0].sourcePullRequest, // assume that all commits are from the same PR
    sourceBranch,
    targetBranch,
    commitMessages,
    PACKAGE_VERSION,
    commits,
  });
}

export function getTitle({
  options,
  commits,
  targetBranch,
}: {
  options: ValidConfigOptions;
  commits: Commit[];
  targetBranch: string;
}) {
  const sourceBranch = getSourceBranchFromCommits(commits);
  const commitMessages = commits
    .map((c) => getFirstLine(c.sourceCommit.message))
    .join(' | ');

  const defaultPrTitle = '[{{targetBranch}}] {{commitMessages}}';

  const template = Handlebars.compile(options.prTitle ?? defaultPrTitle, {
    noEscape: true,
  });

  return template({
    sourcePullRequest: commits[0].sourcePullRequest, // assume that all commits are from the same PR
    sourceBranch,
    targetBranch,
    commitMessages,
    commits,
  }).slice(0, 240);
}

Handlebars.registerHelper('shortSha', getShortSha);
