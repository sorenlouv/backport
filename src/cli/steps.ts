import isEmpty from 'lodash.isempty';
import { setAccessToken } from '../lib/github';

import {
  getCommitsByPrompt,
  getCommitBySha,
  getBranchesByPrompt,
  doBackportVersions,
  handleErrors,
  maybeSetupRepo
} from './cliService';
import { BackportOptions, BranchChoice } from '../types/types';

export async function initSteps(options: BackportOptions) {
  const [owner, repoName] = options.upstream.split('/');
  setAccessToken(options.accessToken);

  try {
    const author = options.all ? null : options.username;
    const commits = options.sha
      ? [await getCommitBySha(owner, repoName, options.sha)]
      : await getCommitsByPrompt(
          owner,
          repoName,
          author,
          options.multipleCommits
        );

    const branches = !isEmpty(options.branches)
      ? (options.branches as string[])
      : await getBranchesByPrompt(
          options.branchChoices as BranchChoice[],
          options.multipleBranches
        );

    await maybeSetupRepo(owner, repoName, options.username);
    await doBackportVersions(
      owner,
      repoName,
      commits,
      branches,
      options.username,
      options.labels
    );
  } catch (e) {
    handleErrors(e);
  }
}
