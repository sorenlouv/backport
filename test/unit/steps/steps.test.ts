import axios from 'axios';
import inquirer from 'inquirer';
import * as createPullRequest from '../../../src/services/github/createPullRequest';
import * as fetchCommitsByAuthor from '../../../src/services/github/fetchCommitsByAuthor';
import * as fs from '../../../src/services/fs-promisified';
import { BackportOptions } from '../../../src/options/options';
import { commitsWithPullRequestsMock } from '../services/github/mocks/commitsByAuthorMock';
import { initSteps } from '../../../src/steps/steps';
import * as childProcess from '../../../src/services/child-process-promisified';

describe('run through steps', () => {
  let rpcExecMock: jest.SpyInstance;
  let rpcExecOriginalMock: jest.SpyInstance;
  let inquirerPromptMock: jest.SpyInstance;
  let axiosHeadSpy: jest.SpyInstance;
  let axiosPostSpy: jest.SpyInstance;

  afterEach(() => {
    jest.clearAllMocks();
  });

  beforeEach(async () => {
    const options: BackportOptions = {
      accessToken: 'myAccessToken',
      all: false,
      apiHostname: 'api.github.com',
      author: 'sqren',
      branches: [],
      branchChoices: [
        { name: '6.x' },
        { name: '6.0' },
        { name: '5.6' },
        { name: '5.5' },
        { name: '5.4' }
      ],
      commitsCount: 10,
      fork: true,
      gitHostname: 'github.com',
      labels: [],
      multiple: false,
      multipleBranches: false,
      multipleCommits: false,
      prDescription: 'myPrDescription',
      prTitle: 'myPrTitle {baseBranch} {commitMessages}',
      repoName: 'kibana',
      repoOwner: 'elastic',
      sha: undefined,
      username: 'sqren'
    };

    rpcExecMock = (childProcess.exec as any) as jest.SpyInstance;
    rpcExecOriginalMock = (childProcess.execAsCallback as any) as jest.SpyInstance;

    jest.spyOn(fs, 'writeFile').mockResolvedValue(undefined);

    jest.spyOn(fetchCommitsByAuthor, 'fetchCommitsByAuthor');
    jest.spyOn(createPullRequest, 'createPullRequest');

    inquirerPromptMock = jest
      .spyOn(inquirer, 'prompt')
      // @ts-ignore
      .mockImplementationOnce(async (args: any) => {
        return { promptResult: args[0].choices[0].value };
      })
      .mockImplementationOnce(async (args: any) => {
        return { promptResult: args[0].choices[0].name };
      });

    // mock verifyAccessToken
    axiosHeadSpy = jest.spyOn(axios, 'head').mockReturnValueOnce(true as any);

    // mock axios post request (graphql)
    axiosPostSpy = jest
      .spyOn(axios, 'post')

      // mock author id
      .mockReturnValueOnce({
        data: {
          data: {
            user: {
              id: 'sqren_author_id'
            }
          }
        }
      } as any)

      // mock list of commits
      .mockReturnValueOnce({
        data: {
          data: commitsWithPullRequestsMock
        }
      } as any)

      // mock create pull request
      .mockReturnValueOnce({
        data: {
          html_url: 'pull request url',
          number: 1337
        }
      } as any);

    await initSteps(options);
  });

  it('should check whether access token is valid', () => {
    expect(axiosHeadSpy).toHaveBeenCalledWith(
      'https://api.github.com/repos/elastic/kibana?access_token=myAccessToken'
    );
  });

  it('should make correct POST requests', () => {
    expect(axiosPostSpy).toMatchSnapshot();
  });

  it('getCommit should be called with correct args', () => {
    expect(fetchCommitsByAuthor.fetchCommitsByAuthor).toHaveBeenCalledWith(
      expect.objectContaining({
        repoName: 'kibana',
        repoOwner: 'elastic',
        username: 'sqren',
        apiHostname: 'api.github.com'
      })
    );
  });

  it('createPullRequest should be called with correct args', () => {
    expect(createPullRequest.createPullRequest).toHaveBeenCalledWith(
      expect.objectContaining({
        repoName: 'kibana',
        repoOwner: 'elastic',
        apiHostname: 'api.github.com'
      }),
      {
        base: '6.x',
        body: `Backports the following commits to 6.x:\n - Add ðŸ‘» (2e63475c)\n\nmyPrDescription`,
        head: 'sqren:backport/6.x/commit-2e63475c',
        title: 'myPrTitle 6.x Add ðŸ‘» (2e63475c)'
      }
    );
  });

  it('prompt calls should match snapshot', () => {
    expect(inquirer.prompt).toHaveBeenCalledTimes(2);
    expect(inquirerPromptMock.mock.calls).toMatchSnapshot();
  });

  it('exec should be called with correct args', () => {
    expect(rpcExecMock).toHaveBeenCalledTimes(9);
    expect(rpcExecMock.mock.calls).toMatchSnapshot();
  });

  it('git clone should be called with correct args', () => {
    expect(rpcExecOriginalMock).toHaveBeenCalledTimes(1);
    expect(rpcExecOriginalMock.mock.calls).toMatchSnapshot();
  });
});
