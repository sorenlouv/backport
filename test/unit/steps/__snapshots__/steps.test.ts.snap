// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`run through steps exec should be called with correct args 1`] = `
Array [
  Array [
    "git remote rm origin",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git remote rm sqren",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git remote add sqren https://myAccessToken@github.com/sqren/kibana.git",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git remote rm elastic",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git remote add elastic https://myAccessToken@github.com/elastic/kibana.git",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git reset --hard && git clean -d --force && git fetch elastic 6.x && git checkout -B backport/6.x/commit-2e63475c elastic/6.x --no-track",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git fetch elastic master:master --force && git cherry-pick 2e63475c483f7844b0f2833bc57fdee32095bacb",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git push sqren backport/6.x/commit-2e63475c:backport/6.x/commit-2e63475c --force",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git checkout master && git branch -D backport/6.x/commit-2e63475c",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
]
`;

exports[`run through steps git clone should be called with correct args 1`] = `
Array [
  Array [
    "git clone https://myAccessToken@github.com/elastic/kibana.git --progress",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
]
`;

exports[`run through steps prompt calls should match snapshot 1`] = `
Array [
  Array [
    Array [
      Object {
        "choices": Array [
          Object {
            "name": "[90m1.[39m Add ðŸ‘» (2e63475c) ",
            "short": "2e63475c",
            "value": Object {
              "existingBackports": Array [],
              "message": "Add ðŸ‘» (2e63475c)",
              "pullNumber": undefined,
              "sha": "2e63475c483f7844b0f2833bc57fdee32095bacb",
            },
          },
          Object {
            "name": "[90m2.[39m Add witch (#85) ",
            "short": "#85",
            "value": Object {
              "existingBackports": Array [],
              "message": "Add witch (#85)",
              "pullNumber": 85,
              "sha": "f3b618b9421fdecdb36862f907afbdd6344b361d",
            },
          },
          Object {
            "name": "[90m3.[39m Add SF mention (#80) [32m6.3[39m",
            "short": "#80",
            "value": Object {
              "existingBackports": Array [
                Object {
                  "branch": "6.3",
                  "state": "MERGED",
                },
              ],
              "message": "Add SF mention (#80)",
              "pullNumber": 80,
              "sha": "79cf18453ec32a4677009dcbab1c9c8c73fc14fe",
            },
          },
          Object {
            "name": "[90m4.[39m Add backport config (3827bbba) ",
            "short": "3827bbba",
            "value": Object {
              "existingBackports": Array [],
              "message": "Add backport config (3827bbba)",
              "pullNumber": undefined,
              "sha": "3827bbbaf39914eda4f02f6940189844375fd097",
            },
          },
          Object {
            "name": "[90m5.[39m Initial commit (5ea0da55) ",
            "short": "5ea0da55",
            "value": Object {
              "existingBackports": Array [],
              "message": "Initial commit (5ea0da55)",
              "pullNumber": undefined,
              "sha": "5ea0da550ac191029459289d67f99ad7d310812b",
            },
          },
        ],
        "message": "Select commit to backport",
        "name": "promptResult",
        "pageSize": 5,
        "type": "list",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "choices": Array [
          Object {
            "name": "6.x",
          },
          Object {
            "name": "6.0",
          },
          Object {
            "name": "5.6",
          },
          Object {
            "name": "5.5",
          },
          Object {
            "name": "5.4",
          },
        ],
        "message": "Select branch to backport to",
        "name": "promptResult",
        "type": "list",
      },
    ],
  ],
]
`;

exports[`run through steps should make correct POST requests 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://api.github.com/graphql",
      Object {
        "query": "
    query getIdByLogin($login: String!) {
      user(login: $login) {
        id
      }
    }
  ",
        "variables": Object {
          "login": "sqren",
        },
      },
      Object {
        "headers": Object {
          "Authorization": "bearer myAccessToken",
          "Content-Type": "application/json",
        },
      },
    ],
    Array [
      "https://api.github.com/graphql",
      Object {
        "query": "
    query getCommitsByAuthorQuery(
      $repoOwner: String!
      $repoName: String!
      $commitsCount: Int!
      $authorId: ID
    ) {
      repository(owner: $repoOwner, name: $repoName) {
        ref(qualifiedName: \\"master\\") {
          target {
            ... on Commit {
              history(first: $commitsCount, author: { id: $authorId }) {
                edges {
                  node {
                    oid
                    message
                    associatedPullRequests(first: 1) {
                      edges {
                        node {
                          number
                          timelineItems(
                            last: 20
                            itemTypes: CROSS_REFERENCED_EVENT
                          ) {
                            edges {
                              node {
                                ... on CrossReferencedEvent {
                                  source {
                                    __typename
                                    ... on PullRequest {
                                      state
                                      baseRefName
                                      commits(first: 20) {
                                        edges {
                                          node {
                                            commit {
                                              message
                                            }
                                          }
                                        }
                                      }
                                    }
                                  }
                                }
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  ",
        "variables": Object {
          "authorId": "sqren_author_id",
          "commitsCount": 10,
          "repoName": "kibana",
          "repoOwner": "elastic",
        },
      },
      Object {
        "headers": Object {
          "Authorization": "bearer myAccessToken",
          "Content-Type": "application/json",
        },
      },
    ],
    Array [
      "https://api.github.com/repos/elastic/kibana/pulls?access_token=myAccessToken",
      Object {
        "base": "6.x",
        "body": "Backports the following commits to 6.x:
 - Add ðŸ‘» (2e63475c)

myPrDescription",
        "head": "sqren:backport/6.x/commit-2e63475c",
        "title": "myPrTitle 6.x Add ðŸ‘» (2e63475c)",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "data": Object {
          "data": Object {
            "user": Object {
              "id": "sqren_author_id",
            },
          },
        },
      },
    },
    Object {
      "type": "return",
      "value": Object {
        "data": Object {
          "data": Object {
            "repository": Object {
              "ref": Object {
                "target": Object {
                  "history": Object {
                    "edges": Array [
                      Object {
                        "node": Object {
                          "associatedPullRequests": Object {
                            "edges": Array [],
                          },
                          "message": "Add ðŸ‘»",
                          "oid": "2e63475c483f7844b0f2833bc57fdee32095bacb",
                        },
                      },
                      Object {
                        "node": Object {
                          "associatedPullRequests": Object {
                            "edges": Array [
                              Object {
                                "node": Object {
                                  "number": 85,
                                  "timelineItems": Object {
                                    "edges": Array [],
                                  },
                                },
                              },
                            ],
                          },
                          "message": "Add witch (#85)",
                          "oid": "f3b618b9421fdecdb36862f907afbdd6344b361d",
                        },
                      },
                      Object {
                        "node": Object {
                          "associatedPullRequests": Object {
                            "edges": Array [
                              Object {
                                "node": Object {
                                  "number": 80,
                                  "timelineItems": Object {
                                    "edges": Array [
                                      Object {
                                        "node": Object {
                                          "source": Object {
                                            "__typename": "PullRequest",
                                            "baseRefName": "6.3",
                                            "commits": Object {
                                              "edges": Array [
                                                Object {
                                                  "node": Object {
                                                    "commit": Object {
                                                      "message": "Add SF mention (#80)

* Add SF mention

* Add several emojis!",
                                                    },
                                                  },
                                                },
                                              ],
                                            },
                                            "state": "MERGED",
                                          },
                                        },
                                      },
                                    ],
                                  },
                                },
                              },
                            ],
                          },
                          "message": "Add SF mention (#80)

* Add SF mention

* Add several emojis!",
                          "oid": "79cf18453ec32a4677009dcbab1c9c8c73fc14fe",
                        },
                      },
                      Object {
                        "node": Object {
                          "associatedPullRequests": Object {
                            "edges": Array [],
                          },
                          "message": "Add backport config",
                          "oid": "3827bbbaf39914eda4f02f6940189844375fd097",
                        },
                      },
                      Object {
                        "node": Object {
                          "associatedPullRequests": Object {
                            "edges": Array [],
                          },
                          "message": "Initial commit",
                          "oid": "5ea0da550ac191029459289d67f99ad7d310812b",
                        },
                      },
                    ],
                  },
                },
              },
            },
          },
        },
      },
    },
    Object {
      "type": "return",
      "value": Object {
        "data": Object {
          "html_url": "pull request url",
          "number": 1337,
        },
      },
    },
  ],
}
`;
